crates/wast/src/ast/expr.rs

```diff
@@ -255,6 +299,51 @@ impl<'a> ExpressionParser<'a> {
         // parse anything else.
         Err(parser.error("too many payloads inside of `(if)`"))
     }
+
+    /// Handles parsing of a `try` statement. A `try` statement is simpler
+    /// than an `if` as the syntactic form is:
+    ///
+    /// ```wat
+    /// (try (do $do) (catch $catch))
+    /// ```
+    ///
+    /// where the `do` and `catch` keywords are mandatory, even for an empty
+    /// $do or $catch.
+    ///
+    /// Returns `true` if the rest of the arm above should be skipped, or
+    /// `false` if we should parse the next item as an instruction (because we
+    /// didn't handle the lparen here).
+    fn handle_try_lparen(&mut self, parser: Parser<'a>) -> Result<bool> {
+        // Only execute the code below if there's a `Try` listed last.
+        let i = match self.stack.last_mut() {
+            Some(Level::Try(i)) => i,
+            _ => return Ok(false),
+        };
+
+        // Try statements must start with a `do` block.
+        if let Try::Do(try_instr) = i {
+            let instr = mem::replace(try_instr, Instruction::End(None));
+            self.instrs.push(instr);
+            if parser.parse::<Option<kw::r#do>>()?.is_some() {
+                *i = Try::Catch;
+                self.stack.push(Level::TryArm);
+                return Ok(true);
+            }
+            return Ok(false);
+        }
+
+        if let Try::Catch = i {
+            self.instrs.push(Instruction::Catch);
+            if parser.parse::<Option<kw::catch>>()?.is_some() {
+                *i = Try::End;
+                self.stack.push(Level::TryArm);
+                return Ok(true);
+            }
+            return Ok(false);
```

Comment:
When this happens (and in `Try::Do` above), the internal state isn't advanced, but I think that might lead to bugs? It's been awhile since I wrote this but in the `If` parsing above each state always advances to the next state unconditionally, even on failed parses I think?

Reply:
Right, I see that `If` does advance the state after each step, but I thought that was because `If` is a bit more permissive than the `Try` case. In particular, if a `do` or `catch` in `try` are missing the intent is to error while for `If` I think it allows the `then` clause to not have the keyword. (or should `try` be more permissive like `if` too?)

My understanding is that if the state isn't advanced, eventually a `Paren::Right` will come up and the `Level::Try(...)` cases there should signal the failing parse. But I might not be understanding the code correctly. :) Happy to try any additional test cases that might clarify this.

Reply:
Ah yeah it's ok to not be as flexible as `if`, although could you expand a bit on the purpose of these then? I'm having a bit of a difficult time wrapping my head around what this is enabling by not returning an error or not updating the state when the group doesn't start with what we'd expect

Reply:
Ah, right. So I think this code could return an error immediately here if the keyword parse fails where it currently returns `Ok(false)`. But I think it would still also need the other check in the `Paren::Right` case, making it a bit redundant (the same error would be returned in both spots, I think).

Specifically, for a case like `(try (do))` with a missing `catch` clause. The state will be advanced to `Try::Catch` after parsing `(do)` but won't go into `handle_try_lparen` again as the expression just ends, so the error should be returned by the right paren parsing.

If you'd prefer I could add the error returns in here too so the control flow is clearer. Or maybe I can add a comment clarifying that it doesn't error yet because the right paren case should catch it anyway?

Reply:
Aha that makes sense, thanks for clarifying! Yeah if you wouldn't mind throwing a comment here that'd be great, and otherwise this looks good to me!

Reply:
Just pushed a new version of the commit with more comments. :)